be/src/storage/primary_key_recover.cpp

```diff
@@ -0,0 +1,96 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/primary_key_recover.h"
+
+#include "serde/column_array_serde.h"
+#include "storage/chunk_helper.h"
+#include "storage/primary_key_encoder.h"
+
+namespace starrocks {
+
+Status PrimaryKeyRecover::recover() {
+    // 1. clean up old primary index and delvec
+    RETURN_IF_ERROR(pre_cleanup());
+    LOG(INFO) << "PrimaryKeyRecover pre clean up finish. tablet_id: " << tablet_id();
+
+    // 2. generate primary key schema
+    std::unique_ptr<Column> pk_column;
+    auto pkey_schema = generate_pkey_schema();
+    if (!PrimaryKeyEncoder::create_column(pkey_schema, &pk_column).ok()) {
+        CHECK(false) << "create column for primary key encoder failed";
+    }
+
+    // For simplicity, we use temp pk index for generate delvec, and then rebuild real pk index when retry publish
+    PrimaryIndex index(pkey_schema);
+
+    OlapReaderStatistics stats;
+    std::vector<uint32_t> rowids;
+    rowids.reserve(DEFAULT_CHUNK_SIZE);
+    PrimaryIndex::DeletesMap new_deletes;
+    auto chunk_shared_ptr = ChunkHelper::new_chunk(pkey_schema, DEFAULT_CHUNK_SIZE);
+    auto chunk = chunk_shared_ptr.get();
+    // 3. scan all rowsets and segments to build primary index
+    RETURN_IF_ERROR(rowset_iterator(
+            pkey_schema, stats,
+            [&](const std::vector<ChunkIteratorPtr>& itrs,
+                const std::vector<std::unique_ptr<RandomAccessFile>>& del_rfs, uint32_t rowset_id) {
+                // handle upsert
+                for (size_t i = 0; i < itrs.size(); i++) {
+                    auto itr = itrs[i].get();
+                    if (itr == nullptr) {
+                        continue;
+                    }
+                    uint32_t rssid = rowset_id + i;
+                    uint32_t row_id_start = 0;
+                    new_deletes[rssid] = {};
+                    while (true) {
+                        chunk->reset();
+                        rowids.clear();
+                        auto st = itr->get_next(chunk, &rowids);
+                        if (st.is_end_of_file()) {
+                            break;
+                        } else if (!st.ok()) {
+                            return st;
+                        } else {
+                            pk_column->reset_column();
+                            PrimaryKeyEncoder::encode(pkey_schema, *chunk, 0, chunk->num_rows(), pk_column.get());
+                            // upsert and generate new deletes
+                            RETURN_IF_ERROR(index.upsert(rssid, row_id_start, *pk_column, &new_deletes));
+                            row_id_start += pk_column->size();
+                        }
+                    }
+                    itr->close();
+                }
+                // handle delete
+                pk_column->reset_column();
+                for (const auto& read_file : del_rfs) {
```

Comment:
Is it correct to process all segments before handling the deletion file? What if upsert and delete files appear alternately?

Reply:
I think we process `upsert` first, and then `delete` when apply?